### ISSUE TITLE
About the training loss

### ISSUE THREAD

**Issue by jkwang93** [ehoogeboom/e3_diffusion_for_molecules#10]
Hi, this is a great job. There are some questions about training, such as how many epochs of loss converge, and what is the final loss of training?

**Comment by ehoogeboom** [comment]
Hi, depends a bit on the dataset, there are some details in Appendix C of the paper. For qm9 we trained for about a week to get to 1100 epochs.

**Comment by jkwang93** [comment]
Thanks for you to answer my questions~
I retrained for thousands of epochs according to your code. Although the valid of the generated molecule is constantly improving to 90%, the nll of the training has not been significantly reduced, and it is always above 2.  And the VAL NLL over one hundred.
Why is this? Is this correct? So how do I judge that I have finished training and got the best model?

**Comment by ehoogeboom** [comment]
The training NLL is actually not the NLL, because the loss is modified the numbers you mention (somewhere above 2) sound reasonable. The validation does compute the actual NLL and starts around the hundreds and then during training usually becomes negative.

The NLL is a little tricky to deal with and not always well-behaved. It's usually best to see with a metric for samples (like the molecule stability) whether the model is working correctly

**Comment by jkwang93** [comment]
Dear Dr.Ehoogeboom，
Thanks a lot for your answer on github, I still have some questions to ask you about the article. Although I have sent you an email, I am worried that the email will be blocked, so I also describe my questions here. You can reply here or by email.

I am a doctoral student whose main research is molecular generation. Recently, I read your paper 'Equivariant Diffusion for Molecule Generation in 3D', which is the first work using the diffusion model in molecular generation, which is a groundbreaking work. I have spent nearly two months researching your article and code with the aim of doing an autoregressive diffusion in 3D molecular generation.

In order to achieve this purpose, I have consulted a lot of information and concluded that there are two ways to achieve autoregressive diffusion:
1. Be inspired by your 'Autoregressive Diffusion Models'. For this paper, if I have not misunderstood it, this article should be adding noise to the characters specified in, for example, NLP in each step. I don't know if my understanding is correct, the idea is also very interesting, and the efficiency is higher.
2. Another idea of ​​autoregressive diffusion is from the paper 'Autoregressive Denoising Diffusion Models for Multivariate Probabilistic Time Series Forecasting', which probably only adds noise to one point at a time, and only sample that point during reconstruction.

For various reasons, I decided to use the second method to implement my autoregressive diffusion method for 3D molecule generation. The overall code is built based on your EDM code, and the entire specific process is:
1. Data set processing: 1) First use minimum_spanning_tree (MST) to traverse the entire molecule; 2) Divide a molecule into t step pairs, each step pair consists of the molecular graph at time t-1 and the atoms that need to be added at time t composition. After this processing, the dataset is dozens of times larger than the original qm9 dataset.
2. Training process: Add noise to the atoms (atomic type + position) that need to be added at time t, add the noised atoms to the molecular graph at time t-1, put them in EGNN for training, and predict the eps value.
3. Sampling process: The sampling process adopts an autoregressive process, sampling one atom at a time. The atoms at time t are sampled at time t, and then added this atom to the molecular graph at time t-1 to obtain the molecular graph at time t.
4. All training without adding charge, because adding charge makes the results worse.

The above is my entire process, but I encountered several problems in the process of implementation:
1. The zero center of gravity: This operation is used in many places in the article and code, and the method of The zero center of gravity is also used each time eps is sampled. I'm thinking, what would go wrong if I didn't use that method?
2. The zero center of gravity in the article is difficult to implement in the method I want to achieve: because only one atom is added at a time, there is no way to use the operation of The zero center of gravity. My approach is to zero center the molecular graph at time t-1, and the atomic coordinates of add at time t are also translated according to the zero center of the molecular graph at time t-1.
3. When sampling, sample one atom at a time, add it to the molecular graph at time t-1, get the molecular graph at time t, and then perform zero center operation on the molecular graph at time t as the next step at time t+1. Molecular graph input.
4. For the problem of The zero center of gravity, I thought of another difficulty in the application: if I want to complete fragment completion molecules in 3D, that is, we known fragments, I need to add atoms to make it a complete molecule. In this task, The zero center of gravity also seems impossible to implement.
5. KL Divergence: I don't know if this item is important. I omitted this item during training to prevent mistakes.

I trained as above and found that valid is always 0. Even when sampling, when zero center is performed on the molecule at time t, because of the type is float32, after zero center, the sum of the positions of all atoms is not 0. 

I don't know what's wrong with my approach, is it a problem with The zero center of gravity or something?

Thank you very much for taking time out of your busy schedule to read my question, and I hope to get your  suggestions~
Sincerely,
Jike Wang

**Comment by jkwang93** [comment]
Dear Dr.Ehoogeboom，
Thanks a lot for your answer on github, I still have some questions to ask you about the article. Although I have sent you an email, I am worried that the email will be blocked, so I also describe my questions here. You can reply here or by email.
I am a doctoral student whose main research is molecular generation. Recently, I read your paper 'Equivariant Diffusion for Molecule Generation in 3D', which is the first work using the diffusion model in molecular generation, which is a groundbreaking work. I have spent nearly two months researching your article and code with the aim of doing an autoregressive diffusion in 3D molecular generation.
In order to achieve this purpose, I have consulted a lot of information and concluded that there are two ways to achieve autoregressive diffusion:
Be inspired by your 'Autoregressive Diffusion Models'. For this paper, if I have not misunderstood it, this article should be adding noise to the characters specified in, for example, NLP in each step. I don't know if my understanding is correct, the idea is also very interesting, and the efficiency is higher.
Another idea of ​​autoregressive diffusion is from the paper 'Autoregressive Denoising Diffusion Models for Multivariate Probabilistic Time Series Forecasting', which probably only adds noise to one point at a time, and only sample that point during reconstruction.
For various reasons, I decided to use the second method to implement my autoregressive diffusion method for 3D molecule generation. The overall code is built based on your EDM code, and the entire specific process is:
Data set processing: 1) First use minimum_spanning_tree (MST) to traverse the entire molecule; 2) Divide a molecule into t step pairs, each step pair consists of the molecular graph at time t-1 and the atoms that need to be added at time t composition. After this processing, the dataset is dozens of times larger than the original qm9 dataset.
Training process: Add noise to the atoms (atomic type + position) that need to be added at

**Comment by jkwang93** [comment]
Dear Dr.Ehoogeboom，
Thanks a lot for your answer on github, I still have some questions to ask you about the article. Although I have sent you an email, I am worried that the email will be blocked, so I also describe my questions here. You can reply here or by email.

I am a PHD. student whose main research is molecular generation. Recently, I read your paper 'Equivariant Diffusion for Molecule Generation in 3D', which is the first work using the diffusion model in molecular generation, which is a groundbreaking work. I have spent nearly two months researching your article and code with the aim of doing an autoregressive diffusion in 3D molecular generation.
In order to achieve this purpose, I have consulted a lot of information and concluded that there are two ways to achieve autoregressive diffusion:
1. Be inspired by your 'Autoregressive Diffusion Models'. For this paper, if I have not misunderstood it, this article should be adding noise to the characters specified in, for example, NLP in each step. I don't know if my understanding is correct, the idea is also very interesting, and the efficiency is higher.
2. Another idea of ​​autoregressive diffusion is from the paper 'Autoregressive Denoising Diffusion Models for Multivariate Probabilistic Time Series Forecasting', which probably only adds noise to one point at a time, and only sample that point during reconstruction.

For various reasons, I decided to use the second method to implement my autoregressive diffusion method for 3D molecule generation. The overall code is built based on your EDM code, and the entire specific process is:
1. Data set processing: 1) First use minimum_spanning_tree (MST) to traverse the entire molecule; 2) Divide a molecule into t step pairs, each step pair consists of the molecular graph at time t-1 and the atoms that need to be added at time t composition. After this processing, the dataset is dozens of times larger than the original qm9 dataset.
2. Training process: Add noise to the atoms (atomic type + position) that need to be added at time t, add the noised atoms to the molecular graph at time t-1, put them in EGNN for training, and predict the eps value.
3. Sampling process: The sampling process adopts an autoregressive process, sampling one atom at a time. The atoms at time t are sampled at time t, and then added this atom to the molecular graph at time t-1 to obtain the molecular graph at time t.
4. All training without adding charge, because adding charge makes the results worse.

The above is my entire process, but I encountered several problems in the process of implementation:
1. The zero center of gravity: This operation is used in many places in the article and code, and the method of The zero center of gravity is also used each time eps is sampled. I'm thinking, what would go wrong if I didn't use that method?
2. The zero center of gravity in the article is difficult to implement in the method I want to achieve: because only one atom is added at a time, there is no way to use the operation of The zero center of gravity. My approach is to zero center the molecular graph at time t-1, and the atomic coordinates of add at time t are also translated according to the zero center of the molecular graph at time t-1.
3. When sampling, sample one atom at a time, add it to the molecular graph at time t-1, get the molecular graph at time t, and then perform zero center operation on the molecular graph at time t as the next step at time t+1. Molecular graph input.
4. For the problem of The zero center of gravity, I thought of another difficulty in the application: if I want to complete fragment completion molecules in 3D, that is, we known fragments, I need to add atoms to make it a complete molecule. In this task, The zero center of gravity also seems impossible to implement.
5. KL Divergence: I don't know if this item is important. I omitted this item during training to prevent mistakes.

I trained as above and found that valid is always 0. Even when sampling, when zero center is performed on the molecule at time t, because of the type is float32, after zero center, the sum of the positions of all atoms is not 0. 

I don't know what's wrong with my approach, is it a problem with The zero center of gravity or something?

Thank you very much for taking time out of your busy schedule to read my question, and I hope to get your  suggestions~
Sincerely,
Jike
